### ISSUE TITLE
Major issue with torque sensing at axis 2 of IIT-Doosan robot

### ISSUE THREAD

**Issue by liesrock** [magician-project/magician-project#4]
Hi @magician-project/unitn ,

We are carrying out many tests with @magician-project/iit-hhcm to provide the rest of the consortium with a stable and working version of the integration of the XBot2 and ROS2 with the Doosan H2515.

A major issue we encountered when **controlling the robot in torque mode** (_torque_rt_ https://manual.doosanrobotics.com/en/api/1.31/Publish/cdrflex-torque_rt) is related to the **reading of the torque read axis 2** of the robot; a simple test to reproduce the issue is the following:

- We put the robot in the configuration reported in the attached picture (all axis in 0 rad except for joint 2 at 1.57 rad) , first without any weight and later on with 10 kg weight attached on the end-effector:
![Doosan_Magician_Issue_Axis_2_torque_reading](https://github.com/user-attachments/assets/60eb65e2-9147-4bf9-8a12-cf5792aea528)

- We read 185 Nm without the weight and 45 Nm with the 10 kg weight (!!!).

Just to give you the full details on the above, the controller used to send the torques to the robot is the following, with kp set at 1200 Nm/rad and kd set at 75 Nm/(rad/sec):

<img width="600" alt="image" src="https://github.com/user-attachments/assets/46b416be-192f-4b42-8c7a-6d0c9a1cc7d8">

The RT data are read using the _read_data_rt_ method https://manual.doosanrobotics.com/en/api/1.31/Publish/cdrflex-read_data_rt, and the torque reading is extrapolated from the _actual_joint_torque_ field of the following struct provided by Dossan API-DRFL https://github.com/doosan-robotics/API-DRFL/blob/e2f558e19bada3cc1541683cba76fefaa683bd30/include/DRFS.h#L850

We also notice that the torque reading would increase if pushing the robot up against the gravity, which is very strange.

Can you replicate the above with the Doosan in Trento so we can understand if there is a problem with the torque sensor reading in our Doosan at IIT or if the issue is for every H2515?
The torque control part looks not affected by this problem which is also strange and the other torque readings in the other axis look good.

**Comment by matteodv99tn** [comment]
Hi,  
We cannot test it today, since the robot is in another facility that we cannot easily reach. But we'll organize ourselves to go there in the next days and provide you some concrete feedback.

Anyway, I can already confirm you that data provided by the APIs are often really strange. Back in September, we tested different torque data, and we settled with [`external_joint_torque`](https://github.com/magician-project/doosan-robot2/blob/30888d3d67408cee65835988e46c56bbc29a5e88/dsr_hardware2/src/dsr_hw_interface2.cpp#L351-L356). The [API](https://github.com/doosan-robotics/API-DRFL/blob/e2f558e19bada3cc1541683cba76fefaa683bd30/include/DRFS.h#L783-L784) is not super-clear on the definition of the different torque measurement (`raw_joint_torque`, `actual_joint_torque`, `external_joint_torque`), but I think everything boils down with the choice of the axis convention that they made (which is quite odd). 
We think that `external_joint_torque` is the overall torque which is applied between the 2 connected links, and is the one you are looking for.

We also found some inconsistency between the controller data and the provided URDF. We tested the [`gravity_torque`](https://github.com/doosan-robotics/API-DRFL/blob/e2f558e19bada3cc1541683cba76fefaa683bd30/include/DRFS.h#L801-L802) data provided by the real time APIs with the one computed by rigid body dynamic libraries (both Pinocchio and Orocos), and the results differs; this implies that either the URDF model or the Doosan controller information are wrong.

Finally, one last inconsistency that we found is the 2nd axis compensation. Looking at the [documentation of the `torque_rt`](https://manual.doosanrobotics.com/en/api/1.31/Publish/cdrflex-torque_rt) function, they state that for H Series robot (like ours) the second axis is gravity compensated. Still, the last time we tested the [ROS2 hardware interface](https://github.com/magician-project/doosan-robot2/blob/30888d3d67408cee65835988e46c56bbc29a5e88/dsr_hardware2/src/dsr_hw_interface2.cpp#L387-L394) we modified, we had to manually compensate in gravity all axis. I don't know if you also noticed this behavior.

**Comment by matteodv99tn** [comment]
Hi @liesrock, this morning we went to our Facility and did some testing. We did not have that much time to get some proper weights (like the 10kg you have), but we did some qualitative testing _by hand_.

From our view-point, these are joint axis directions over which the position is calculated. In this case, the robot is in a configuration _(0, -20°, -70°, 0, 0, 0)_.
![robot-axis](https://github.com/user-attachments/assets/2df87d23-58e7-4ddc-8fda-8fb05fb17ea8)

Here you may find the video where @SwonGao first applies an upward force, and then a downward force.

https://github.com/user-attachments/assets/92b9b651-caaa-446f-b63f-0a9259ba21e8

The following is the plot of the torque measured at the second and third joint for the `actual_joint_torque` and `external_joint_torque` data (_actual_joint1_, _actual_joint2_, and _external_joint1_, _external_join2_ respectively). 
For both axis, retrieved values present a reasonable behavior.

![image](https://github.com/user-attachments/assets/9bdf692b-f05d-4fce-845d-319ceb5a2be3)

We also logged `raw_joint_torque` data, which appears to be (at a macroscopic level) the same as the `actual_joint_torque`.

For sake of completeness, we performed the same test in the dual configuration, with joints at _(0, 20°, 70°, 0, 0, 0)_. Attached is the video and the plot of the collected data, which also appears to be _consistent with physics laws_:


https://github.com/user-attachments/assets/264ba2c3-a063-4f0f-8611-651a30b702c5

![image](https://github.com/user-attachments/assets/9387e13d-985c-43fe-b0c0-29b5d9f66454)

Data have been plotted (and stored) using [rerun.io](https://rerun.io/) v 0.20.3, so if you want I can share the database with you, so that you can check at all data.

We might plan to do further testing next monday, so if you want some more specific information or diverse data, let us know in advance so that we prepare ourselves.

**Comment by liesrock** [comment]
@matteodv99tn @SwonGao, thanks a lot for your support and the tests: the behavior seems different compared to our Doosan.

On our side, we will try a similar test, just reading the torque through ROS and avoiding the RT communication with the robot; on your side, can you replicate our test controlling the robot using the torque_rt function and having the joint configuration that we had? We can then compare the torque readings of the two robots in the same condition.

Thanks again!

**Comment by matteodv99tn** [comment]
Yes, on Monday me and @SwonGao will prepare some code and will test everything within our facility, and will try to use an joint-level impedance control scheme using the `torque_rt` function. We'll also use proper weights, so we can compare results at a quantitative level. We'll share with you the code as well.

Due to the currently unstable mounting configuration from our side, we cannot guarantee to perform the test with the (0, 90°, 0, 0, 0, 0) configuration you asked, but more details will follow!

**Comment by EdoardoRomiti** [comment]
Hi @matteodv99tn @SwonGao thanks for your tests, very helpful. 
I just wanted to add that it does indeed look like a reasonable behavior from your experiments, but the absolute value of the torques is still weird in my opinion. We see that the abs value of the torque of joint 1 (first pitch joint) is smaller than the one of joint 2 (second pitch joint)! This is not possible in both configurations you considered unless I'm missing something obvious.

With @liesrock we also noticed (as you did) that torque readings and torques computed by rigid body dynamics libraries (RBDL, Pinocchio, KDL) differ. We also thought that it might be due to a discrepancy between URDF and real robot. But maybe it's not the case and the dynamic libraries are actually correct. It seems from both ours and yours experiments that the second joint of the doosan reads a torque which is not correct. Maybe it can be that the gravity compensator mentioned in the warning [here](https://manual.doosanrobotics.com/en/api/1.31/Publish/cdrflex-torque_rt) has some effect on the readings as well?

If for instance this compensator is not software but actually a passive mechanical system its torque would be not be measured by the torque sensor and could explain these readings and also some of the issues we noticed. Still it would not explain the behaviour we saw when adding the weight. 

As @liesrock said we will run your same test and probably also repeat the one with weights, so hopefully we understand better where is the issue.

**Comment by matteodv99tn** [comment]
As you mentioned, we also think that the _gravity compensator on the second axis_ mentioned in the documentation is actually a mechanical component, whose torque contribution is not measured by the torque sensor. This also justifies the mismatch between the sensor readings and the results from the rigid body dynamic libraries.

Concerning the "strange" reading at the shoulder joint, we confirm that we don't have the issue. In the configuration (0, 20°, 70°, 0, 0, 0), we measure:

- -53N/m at shoulder, -70N/m at the elbow, with no payload;
- -93N/m at shoulder, -95N/m at the elbow, with a payload of approx. 4kg.

I wanted to provide you more detailed results, but due to my inexperience with the rerun.io sdk which I started using little time ago, the data I collected are not stored nicely to do some post-processing; I'm working on creating some plots, but if you want to see directly the sensor reading, you may find [here](https://drive.google.com/drive/folders/1Jh1SxhIEnwt4lJ2WUBeOIdQHh1jQw5J4?usp=drive_link) the rerun recordings. Files are named with the format `<shoulder deg> <elbow deg> <payload> <joint stiffness>`.

The source code for the code we run can be found [here](https://github.com/magician-project/doosan-torque-testing/blob/main/executables/main_test.cpp), but essentially we reach a desired joint configuration, and then we record 10secs of samples while controlling the robot in joint impedance control with the `torque_rt` function.
